Support for CDT / Xilinx Vitis < 2023.1

Arryk opened the pull request:
Basic support for CDT with an initial port of the themes. 

**Notes:**
- Some of the styling options are different from Java (Preprocessor, macro etc...) so I had to make some educated choices on the design.
- The themes  port are not perfect, pretty sure they could be tweaked a little bit for a better feel relative to its JDT equivalent
- I tried to create a color definition in Neptune to try to reuse color and prevent duplicating RGB raw values into IEclipsePreferences but this failed, if you know a way of doing it that'll make the tweaking easier

I have successfully exported the plugin 0.3.0 and installed it on Xilinx Vitis 2021.1

Arryk committed:
Neptune #1
> Added relevant preferences for CDT in planets.css and nepture.css
> Fixed typo in moon.css
> Changed manifest to JavaSE-11
> Added Vitis as a target

Arryk committed:
Merge branch 'PyvesB:master' into master

Arryk committed:
Added pluto theme, some

Arryk committed:
Merge branch 'master' of github.com:Arryk/eclipse-planet-themes-vitis

Arryk committed:
Updated target platform to locate Vitis for debug

PyvesB commented on planet-themes-targetplatform/planet-themes-targetplatform.target:
```diff
@@ -15,5 +15,6 @@
 			<unit id="org.eclipse.osgi" version="0.0.0"/>
 			<unit id="com.google.gson" version="0.0.0"/>
 		</location>
+		<location path="C:\Xilinx\Vitis\2021.1\eclipse\win64.o" type="Profile"/>
```
Folks who do not have this element in their system will not be able to resolve the target platform. It's fine to add it as a local change to test things out, but we really shouldn't push it to the repository. :)

Arryk replied:
Will get removed in next commit

PyvesB commented on planet-themes-plugin/css/moon.css:
```diff
@@ -39,4 +39,4 @@ IEclipsePreferences#org-eclipse-debug-ui:io-github-pyvesb-eclipse_planet_themes-
 IEclipsePreferences#org-eclipse-jdt-ui:io-github-pyvesb-eclipse_planet_themes-moon {
   preferences: 'sourceHoverBackgroundColor=241,241,241';
-}
+}
```
Is this line ending change intentional?

Arryk replied:
I do not have this line change in my local timeline for this file, I suppose it is an artifact of synchronizing the fork

This is my local version, I just removed empty line 43 and changed neptune to moon
<img width="834" alt="image" src="https://github.com/PyvesB/eclipse-planet-themes/assets/12896037/ce81abf4-301c-495a-9a11-c4821bfb46d3">

looks like the changes are a noop

PyvesB commented on planet-themes-plugin/css/neptune.css:
```diff
@@ -162,3 +163,126 @@ IEclipsePreferences#org-eclipse-pde-ui:io-github-pyvesb-eclipse_planet_themes-ne
+    'org.eclipse.cdt.internal.ui.text.doctools.single         =	196, 217, 240' 	/* COMMENT */
+}
+
+/* Not necessary at the moment
```
Let's remove it then! Besides these values are already defined here: https://github.com/PyvesB/eclipse-planet-themes/blob/248fd762f8141f0fbd7ce152859d067581f335dd/planet-themes-plugin/css/planets-dark.css#L32

Arryk replied:
Removed on next commit

PyvesB commented on planet-themes-plugin/css/neptune.css:
```diff
@@ -162,3 +163,126 @@ IEclipsePreferences#org-eclipse-pde-ui:io-github-pyvesb-eclipse_planet_themes-ne
+    'pp_default                                               = 38,  144, 219' 	/* DEFAULT */ 
+    'pp_directive                                             = 105, 125, 213' 	/* PREPROC */
+    'pp_header                                                = 84,  120, 234' 	/* STRING */
+    'semanticHighlighting.annotation.color                    = 196, 181, 184'
+    'semanticHighlighting.number.color                        = 104, 151, 187' 	/* NUMBERS */
+    'semanticHighlighting.recordKeyword.color                 = 105, 159, 213' 	/* KEYWORD */
+    'semanticHighlighting.restrictedKeywords.color            = 105, 159, 213'	/* KEYWORD */
+    'semanticHighlighting.varKeyword.color                    = 105, 159, 213' 	/* KEYWORD */
+    'semanticHighlighting.yieldKeyword.color                  = 121, 171, 255'  /* TYPEDEF */
```
These feel JDT-specific, are they relevant in CDT?

Arryk replied:
You are right, they have no effect, removed

PyvesB commented on planet-themes-plugin/css/pluto.css:
```diff
@@ -162,3 +162,56 @@ IEclipsePreferences#org-eclipse-pde-ui:io-github-pyvesb-eclipse_planet_themes-pl
+    'pp_default                                               = 236, 154, 24'
+    'pp_directive                                             = 189, 51,  30'
+    'pp_header                                                = 189, 51,  30'
+    'semanticHighlighting.annotation.color                    = 189, 51,  30'
+    'semanticHighlighting.number.color                        = 189, 51,  30'
```
Same as for Neptune, these feel JDT-specific, are they relevant in CDT?

Arryk replied:
They are not, removed

PyvesB commented on planet-themes-plugin/css/neptune.css:
```diff
@@ -162,3 +163,126 @@ IEclipsePreferences#org-eclipse-pde-ui:io-github-pyvesb-eclipse_planet_themes-ne
 }
+
+
+/* Color Palette (This is informative, there is no way of interpolating this into preference tag) */
```
If we don't find a way to make these work, I'd suggest removing them, they'll just be more things to load by Eclipse at startup. :)

Arryk replied:
After discussing the matter with vogella, this will be removed in the next commit

PyvesB commented:
> I tried to create a color definition in Neptune to try to reuse color and prevent duplicating RGB raw values into IEclipsePreferences but this failed, if you know a way of doing it that'll make the tweaking easier

@vogella you're one of the experts when it comes to CSS, do you happen to know if this is possible? :)

PyvesB reviewed (commented):
Thanks for the contribution, I've given this a spin, and overall this looks great! I like how you did the indentation on the newly-added sections. 👍🏻 

I've left a few comments inline, but overall nothing major, mostly things that I think can be removed. However, one thing we should probably add is the following section:
````
IEclipsePreferences#org-eclipse-ui-editors:org-eclipse-cdt-ui {
	preferences:
		"org.eclipse.cdt.ui.occurrenceIndicationColor=xxxxx"
		"org.eclipse.cdt.ui.writeOccurrenceIndicationColor=yyyyyy"
}
````

Without it, it can be very challenging to read text when occurrences are marked:

![Screenshot 2024-06-25 224630](https://github.com/PyvesB/eclipse-planet-themes/assets/10694593/e7b26c69-e0b9-4888-8d4c-b3aa5eed7348)

We can simply align on the same color values used for other occurrenceIndication properties. :)

vogella commented:
> > I tried to create a color definition in Neptune to try to reuse color and prevent duplicating RGB raw values into IEclipsePreferences but this failed, if you know a way of doing it that'll make the tweaking easier
> 
> @vogella you're one of the experts when it comes to CSS, do you happen to know if this is possible? :)

I don't understand the question. You can define colors via extension point and re-use this definition in CSS. Was that the question?

Arryk commented:
> > > I tried to create a color definition in Neptune to try to reuse color and prevent duplicating RGB raw values into IEclipsePreferences but this failed, if you know a way of doing it that'll make the tweaking easier
> > 
> > 
> > @vogella you're one of the experts when it comes to CSS, do you happen to know if this is possible? :)
> 
> I don't understand the question. You can define colors via extension point and re-use this definition in CSS. Was that the question?

Hi @vogella 

The goal here is to reuse ColorDefinition inside IEclipsePreference `preferences:` property.
Image as an example:
<img width="403" alt="image" src="https://github.com/PyvesB/eclipse-planet-themes/assets/12896037/df5a9655-2781-4bba-8cd2-9edbcf81586e">

The `preference:` property is string that equals a raw value of RGB, is there a way to replace that raw RGB value with a value inferred from ColorDefinition ?

For example:
```
IEclipsePreferences#org-eclipse-ui-editors:io-github-pyvesb-eclipse_planet_themes-moon {
  preferences:
    'AbstractTextEditor.Color.Background=#io-github-pyvesb-eclipse_planet_themes-BACKGROUND_COLOR'
    'AbstractTextEditor.Color.SelectionBackground=190,190,190'
    'currentLineColor=220,220,220'
    'searchResultIndicationColor=128,128,128';
}
```
but this doesn't seems to work, any way we can achieve this result ? 
Can we access a property like `AbstractTextEditor.Color.Background` and set it to `#io-github-pyvesb-eclipse_planet_themes-BACKGROUND_COLOR`

Hope this clarify

vogella commented:
I don't think that will work. What will work is that you use the colors extension point to define a color. That will make the color also available in the colors and themes preference page so that the user can change it. This color can also be used in the CSS. I don't think that it would be possible to use this in the preference node at the moment, IIRC this copies the stuff as is.

But changing the CSS preference handler should be possible, maybe build in your use case. IIRC it is super simple at the moment https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.e4.ui.css.swt/src/org/eclipse/e4/ui/css/swt/properties/preference/EclipsePreferencesHandler.java and https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/bundles/org.eclipse.e4.ui.css.swt/src/org/eclipse/e4/ui/css/swt/helpers/EclipsePreferencesHelper.java look relevant. 

Also the relevant tests are here https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/tests/org.eclipse.e4.ui.tests.css.swt/src/org/eclipse/e4/ui/css/swt/properties/preference/EclipsePreferencesHandlerTest.java

Arryk commented:
Thanks for the quick answer, this is beyond my comprehension of Java and Eclipse. It looks to me like a lot of work for not much. 
I suppose writing a Python or Julia script to automatically interpolates the CSS file directly would be easier and more elegant, but it is still a lot of work for not much results.

Arryk committed:
netup.css > Removed unecessary items, added write occurance color to CDT
pluto.css > Removed unecessary items, added write occurance color to CDT
.target > Removed Vitis location

Arryk commented:
> Thanks for the contribution, I've given this a spin, and overall this looks great! I like how you did the indentation on the newly-added sections. 👍🏻
> 
> I've left a few comments inline, but overall nothing major, mostly things that I think can be removed. However, one thing we should probably add is the following section:
> 
> ```
> IEclipsePreferences#org-eclipse-ui-editors:org-eclipse-cdt-ui {
> 	preferences:
> 		"org.eclipse.cdt.ui.occurrenceIndicationColor=xxxxx"
> 		"org.eclipse.cdt.ui.writeOccurrenceIndicationColor=yyyyyy"
> }
> ```
> 
> Without it, it can be very challenging to read text when occurrences are marked:
> 
> ![Screenshot 2024-06-25 224630](https://private-user-images.githubusercontent.com/10694593/342924094-e7b26c69-e0b9-4888-8d4c-b3aa5eed7348.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkzODk3NzAsIm5iZiI6MTcxOTM4OTQ3MCwicGF0aCI6Ii8xMDY5NDU5My8zNDI5MjQwOTQtZTdiMjZjNjktZTBiOS00ODg4LThkNGMtYjNhYTVlZWQ3MzQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI2VDA4MTExMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5OGFkMDQ2MjM2YmMxNTEyMzBkODFlNjM4MzFhZmY5ZDIxMzJiZGQ3ZmQwZGM0Y2JhZmQ4ZTlmMjVhYzc5OGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.o1VU0lWuYX1gXxMBc4mHa8VvOswOOqIWCIfwfIjWkIQ)
> 
> We can simply align on the same color values used for other occurrenceIndication properties. :)

Commit with fixes available.
I added the write occurrences color but it didn't seems to work, can you test on your side ?
Thanks

Arryk committed:
pluto.css > Tweaked some enum and preprocessor color to have better feel for C

Arryk committed:
pluto.css > Fixed color issue with external SDK color being blue

Arryk commented:
I played a little bit with the Pluto theme in a real environment and couple of things don't feel right, custom types are not very readable whilst native types contrast is perfect

<img width="184" alt="image" src="https://github.com/PyvesB/eclipse-planet-themes/assets/12896037/2dff8964-7a29-47e7-8145-1231f82a7778">

Typedefs being the same color as their field is not very readable, any idea what color to attribute to typedefs and custom types ?


<img width="138" alt="image" src="https://github.com/PyvesB/eclipse-planet-themes/assets/12896037/7de509ea-5001-498c-ae5d-bd71dba927d2">


when it comes to ENUM I already tweaked a little bit the color for the fields to improve in-code readability

PyvesB commented:
>  I added the write occurrences color but it didn't seems to work, can you test on your side ?

They need to be put in a different preference section:
````
IEclipsePreferences#org-eclipse-ui-editors:org-eclipse-cdt-ui { ...
````

> Typedefs being the same color as their field is not very readable, any idea what color to attribute to typedefs and custom types ?

Sorry, I've not written C in years, which bits are problematic in above screenshots?

Arryk committed:
Moved occurence indication to the right preference section

Arryk commented:
>They need to be put in a different preference section:

Ok, after doing the fix it now works 👍 

> Sorry, I've not written C in years, which bits are problematic in above screenshots?

I have included some screenshots in my previous message, you can see that My_Struct struct1; Union and uint8_t are all similar color, in bigger project and struct this is not very readable, everything is yellow. 

I can always play with the settings and modify it, just wanted to see if you have an opinion on the matter. Maybe this is looking as intended

PyvesB commented:
For reference, here are how things look in the default CDT dark-theme:
<img width="246" alt="Screenshot 2024-06-30 at 18 16 38" src="https://github.com/PyvesB/eclipse-planet-themes/assets/10694593/5934b8e5-33c3-4cef-9a1b-d0946d8e5d8b">

Overall, I've got no strong opinions on the matter, I'd just suggest not extending the colour palette too much. :)

Arryk commented:
Okay, let's just go with the current state of things

PyvesB commented:
Sounds good, we can always improve later on. This is great, thanks so much for your contribution!
